spring-integration-debezium/src/main/java/org/springframework/integration/debezium/inbound/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,289 @@
+	}
+
+	/**
+	 * Enables support for tombstone (aka delete) messages. On a database row delete, Debezium can send a tombstone
+	 * change event that has the same key as the deleted row and a value of {@link Optional.empty}. This record is a
+	 * marker for downstream processors. It indicates that log compaction can remove all records that have this key.
+	 *
+	 * When the tombstone functionality is enabled in the Debezium connector configuration you should enable the empty
+	 * payload as well.
+	 */
+	public void setEnableEmptyPayload(boolean enableEmptyPayload) {
+		this.enableEmptyPayload = enableEmptyPayload;
+	}
+
+	/**
+	 * Set the {@link ExecutorService}, where is not provided then a default of single thread Executor will be used.
+	 * @param executorService the executor service.
+	 */
+	public void setExecutorService(ExecutorService executorService) {
+		this.executorService = executorService;
+		this.executorServiceExplicitlySet = true;
+	}
+
+	/**
+	 * Outbound message content type. Should be aligned with the {@link SerializationFormat} configured for the
+	 * {@link DebeziumEngine}.
+	 */
+	public void setContentType(String contentType) {
+		this.contentType = contentType;
+	}
+
+	/**
+	 * Specifies how to convert Debezium change event headers into Message headers.
+	 *
+	 * @param headerMapper concrete HeaderMapping implementation.
+	 */
+	public void setHeaderMapper(HeaderMapper<List<Header<Object>>> headerMapper) {
+		Assert.notNull(headerMapper, "'headerMapper' must not be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "debezium:inbound-channel-adapter";
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+
+		Assert.notNull(this.executorService, "Invalid Executor Service. ");
+		Assert.notNull(this.headerMapper, "Header mapper can not be null!");
+
+		if (!this.enableBatch) {
+			this.debeziumEngineBuilder.notifying(new StreamChangeEventConsumer<byte[]>());
+		}
+		else {
+			this.debeziumEngineBuilder.notifying(new BatchChangeEventConsumer<byte[]>());
+		}
+
+		this.debeziumEngine = this.debeziumEngineBuilder.build();
+	}
+
+	@Override
+	protected void doStart() {
+		if (this.latch.getCount() > 0) {
+			return;
+		}
+		this.latch = new CountDownLatch(1);
+		this.future = this.executorService.submit(() -> {
+			try {
+				this.debeziumEngine.run();
+			}
+			finally {
+				this.latch.countDown();
+			}
+		});
+	}
+
+	@Override
+	protected void doStop() {
+		if (this.future.isDone()) {
+			return;
+		}
+		this.future.cancel(true);
+		try {
+			if (!this.latch.await(5, TimeUnit.SECONDS)) {
+				throw new IllegalStateException("Failed to stop " + this);
+			}
+		}
+		catch (InterruptedException ignored) {
+		}
+	}
+
+	@Override
+	public void destroy() {
+		super.destroy();
+		if (!this.executorServiceExplicitlySet) {
+			this.executorService.shutdown();
+		}
+		try {
+			this.executorService.awaitTermination(10, TimeUnit.SECONDS);
```

Comment:
This is not OK for externally provided executor.
We really cannot control how that one is shouted down.
We may wait on an internal one, but why